projects/RabbitMQ.Client/client/api/IEndpointResolverExtensions.cs

```diff
@@ -71,12 +71,12 @@ public static class EndpointResolverExtensions
                 }
             }
 
-            if (EqualityComparer<T>.Default.Equals(t, default(T)) && exceptions.Count > 0)
+            if (EqualityComparer<T>.Default.Equals(t!, default!) && exceptions.Count > 0)
             {
                 throw new AggregateException(exceptions);
             }
 
-            return t;
+            return t!;
```

Comment:
The null-forgiving operator is incorrect here. `t` technically can be null here.

Instead of this, it's better to change the return type from `Task<T>` to `Task<T?>`.

Reply:
Yes that's true, unfortunately if we'd change the annotation, then that'd mean a whole lot of null checks. So I'd rather throw an exception if there are no endpoints in the resolver. @lukebakken your take

Reply:
Yes, if the `selector` returns `null` there is a serious issue.

Reply:
> Yes, if the `selector` returns `null` there is a serious issue.

Where is the serious issue? If it's in this method, this method should throw an exception instead of lying that it is returning an value when declaring to only return non-null.

That will cause a `NullReferenceException` to be thrown with no obvious reason to the caller.

Reply:
@lukebakken this comment is marked as resolved, although participants agree that logic here should be changed to throw an exception in case of null.

Reply:
@lukebakken Should we therefore throw here unconditionally if we get default(T) back? So we can avoid changing the return type?

Reply:
See 6616c4fbac85f98ba0a1bfdc048dee2c7da6bcdc